[방탈출 결제/배포 STEP1] 노랑(노주희) 미션 제출합니다.

yesjuhee opened the pull request:
안녕하세요 범블비! 리뷰이 노랑입니다 🙌 

[이번 미션의 변경 범위입니다!](https://github.com/woowacourse/spring-roomescape-payment/pull/236/files/6118c93fba51ee05e456a083d996f8ca2bcdb29c..e3ff2e23327637191292be7d6131c63224e2c4e4)
리뷰 잘 부탁드립니다 😄 

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  https://prolog.techcourse.co.kr/studylogs/4477

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- [x]  한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
- [x]  else 예약어를 쓰지 않는다.
- [x]  모든 원시 값과 문자열을 포장한다.
- [ ]  일급 컬렉션을 쓴다.
- [x]  한 줄에 점을 하나만 찍는다.
- [x]  줄여 쓰지 않는다(축약 금지).
- [x]  모든 엔티티를 작게 유지한다.
- [ ]  3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
- [x]  getter/setter/프로퍼티를 쓰지 않는다.

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

코드에 적겠습니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

rladmstn committed:
chore: 이전 코드 마이그레이션

rladmstn committed:
feat: `TossPayments` 결제 위젯 연동 및 UI/스타일 업데이트

- 예약 결제 기능에 `TossPayments` 결제 위젯 추가
- `reservation.html`에 결제 관련 UI 구성 요소 및 스크립트 추가
- `toss-style.css` 생성 및 결제 위젯 관련 스타일 정의
- `user-reservation.js`에 결제 위젯 초기화 및 처리 로직 추가

rladmstn committed:
docs: 예약 추가 API 문서에 결제 내용 추가

- README에 대기 API 항목 추가
- 예약 API에 결제 정보 필드(`paymentKey`, `orderId`, `amount`) 추가
- 예약 API 시나리오 및 응답 상세 설명 추가

rladmstn committed:
feat: 예약 엔티티에 결제 도메인 추가

- Reservation에 Payment 필드 추가 및 단방향 연결 (@OneToOne)
- Amount 임베디드 타입 생성 (결제 금액 값 관리)
- PaymentKey 임베디드 타입 생성 (결제 키 관리)
- Payment 엔티티 생성 (PaymentKey와 Amount 포함)

rladmstn committed:
feat: 결제 정보를 포함한 예약 기능 추가

- 결제 정보 저장을 위한 `Payment` 엔티티 및 `Repository` 추가
- `Reservation` 생성 시 결제 정보 연동 로직 추가
- `@Transactional` 어노테이션 도입으로 예약 생성 트랜잭션 처리

rladmstn committed:
refactor: 결제 성공 알림 메시지 추가

- 예약 결제 성공 시 사용자에게 알림 메시지 표시

rladmstn committed:
feat: `TossPayments` 예외 처리 로직 구현

- `TossPayments` 관련 DTO 패키지 변경 및 TossErrorResponse 클래스 추가
- `TossPaymentsClient`에 `secretKey` 의존성 주입 및 예외 처리 로직 구현
- `TossPaymentsException` 커스텀 예외 추가 및 `GlobalExceptionHandler`에서 처리 로직 추가

rladmstn committed:
refactor: 결제 로직 리팩토링 및 패키지 구조 개선

- `Payment` 관련 로직을 `PaymentService`로 분리하여 `ReservationCommandService` 간소화
- 결제 관련 도메인과 레포지토리를 `reservation` 패키지에서 `payment` 패키지로 이동

rladmstn committed:
feat: `TossPaymentsClient` 리팩토링 및 타임아웃 예외 처리 추가

- `ClientConfig`에서 HTTP 요청 타임아웃 설정 및 Authorization 헤더 처리 추가
- `ResourceAccessException`을 처리하는 시간 초과 예외 핸들러 추가

rladmstn committed:
test: 결제 API 테스트 추가

rladmstn committed:
test: `SpringBootTest`에 랜덤 포트 적용

rladmstn committed:
fix: API 요청 형식 변경 사항 반영

yesjuhee commented on src/main/java/roomescape/client/TossPaymentsClient.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private void handlerTossPaymentsException(ClientHttpResponse res) throws IOException {
+        final ObjectMapper objectMapper = new ObjectMapper().configure(
+                DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        final TossErrorResponse errorResponse = objectMapper.readValue(res.getBody(),
+                TossErrorResponse.class);
```
[결제 승인 API 호출 실패](https://docs.tosspayments.com/reference/error-codes#%EA%B2%B0%EC%A0%9C-%EC%8A%B9%EC%9D%B8)시 아래와 같이 `code`와 `message` 필드가 전송됩니다. 이때 `code`는 지금 프로젝트에서 활용할 만 한 부분이 없을 것 같아서 `message`만 가지는 `TossErrorResponse` DTO를 정의해서 사용했습니다!

```json
{
  "code": "NOT_FOUND_PAYMENT",
  "message": "존재하지 않는 결제 입니다."
}
```

ddaaac replied:
당장 필요없어보여도 필드 정의 정도는 해두면 나중에 쓸 수도 있을 것 같아요. 에러 상황에 대한 로그를 남길 때도 유용하게 사용할 수 있을 것 같구요.

yesjuhee replied:
오호 에러 로그는 생각해보지 못했는데 좋은 관점이네요.
반영했습니다!

28963b91

yesjuhee commented on src/main/java/roomescape/global/exception/GlobalExceptionHandler.java:
```diff
@@ -92,6 +102,18 @@ public ErrorResponse handleIllegalArgumentException(final IllegalArgumentExcepti
         return new ErrorResponse(e.getMessage());
     }
 
+    @ExceptionHandler(ResourceAccessException.class)
+    public ResponseEntity<ErrorResponse> handleSocketTimeoutException(final ResourceAccessException e) {
+        Throwable cause = e.getCause();
+        while (cause != null) {
+            if (cause instanceof SocketTimeoutException) {
+                return ResponseEntity.internalServerError().body(new ErrorResponse("시간이 초과되었습니다. 다시 시도해 주세요."));
+            }
+            cause = cause.getCause();
+        }
+        return ResponseEntity.internalServerError().body(new ErrorResponse("알 수 없는 에러가 발생했습니다."));
+    }
```
`ClientConfig` 클래스에 설정해 놓은 TimeOut 시간을 초과할 경우 발생하는 `ResourceAccessException`에 대한 핸들러입니다. 이 때 이 예외의 `cause` 중 `SocketTimeoutException`이 발생한 경우에만 시간 초과에 대한 메시지를 반환하도록 했습니다.

ddaaac replied:
잘 잡아주셨네요.

yesjuhee commented on src/main/java/roomescape/global/exception/GlobalExceptionHandler.java:
```diff
@@ -1,23 +1,33 @@
 @RestControllerAdvice
 public class GlobalExceptionHandler {
 
+    @ExceptionHandler(TossPaymentsException.class)
+    public ResponseEntity<ErrorResponse> handleTossPaymentsException(final TossPaymentsException e) {
+        return ResponseEntity.status(e.getStatusCode().value())
+                .body(new ErrorResponse(e.getMessage()));
+    }
```
토스에서 [에러 응답](https://docs.tosspayments.com/reference/error-codes#%EA%B2%B0%EC%A0%9C-%EC%8A%B9%EC%9D%B8)을 보내줄 경우에는 그 상태코드와 메시지를 그대로 사용했습니다.
이미 정의가 잘 되어있어서 가공할 필요를 느끼지 못했습니다ㅎㅎ

ddaaac replied:
토스에서 반환하는 메세지를 그대로 클라이언트에 보여주는 게 안전할지 고민해보시면 좋을 것 같아요.

yesjuhee replied:
에러 메시지를 검토했을 때 보안에 취약할 만 한 지점은 발견하지 못했는데요
일부 메시지는 일반 사용자가 봤을 때 이해할 수 없는 내용이라 사용자 경험에 좋지 않을 것이라 판단했습니다.
따라서 특정 에러 코드의 경우 토스 에러 메시지를 넘기지 않고 "결제에 실패했습니다. 잠시 후 다시 시도해주세요." 라고 메시지를 반환하도록 설정했습니다.

65767928

다만 이 부분에서 우려되는 점이 있습니다. 이러한 처리 방식은 **토스 API에 너무 의존적**이지 않나라는 생각이 들기도 합니다. 토스가 에러 코드를 추가하거나 수정하면 방탈출 예약의 코드도 같이 변경이 되어야 하는 구조입니다. 이런 현상에 대해서는 토스 API가 변경될 일이 거의 없다고 가정하고 그냥 개발하는게 좋을까요?

ddaaac replied:
맞습니다. 현재 방식은 외부 API에 너무 의존적이에요.

그래서 외부 API를 쓸 때는 우리가 신경써야할 코드들을 모아서 따로 관리하고 유저에게 메세지를 응답하고, 나머지 코드는 로그 정도만 남기고 유저에게는 "서버 에러가 발생했습니다" 정도로 노출하는 쪽이 관리에 유용합니다.
꼭 신경써야할 에러 코드들만 관리해보면 좋을 것 같아요.

yesjuhee commented on src/main/java/roomescape/payment/domain/Payment.java:
```diff
@@ -0,0 +1,45 @@
+@Entity
+public class Payment {
+
+    @Id
+    private PaymentKey paymentKey;
```
토스에서 보내주는 `paymentKey`가 곧 결제의 식별자라고 생각해서 AUTO_INCREMENT id를 사용하지 않았습니다.
`paymentKey`가 있어서 별도의 식별자가 불필요하다고 생각했는데 혹시 범블비는 어떻게 생각하시나욥?

ddaaac replied:
개인적으로는 외부 시스템은 믿지 않는 쪽이 맞다고 생각해서 자체 키를 둘 것 같긴합니다.
굳이 외부 시스템의 고유함을 믿을 필요는 없어보여요.

yesjuhee replied:
토스의 고유한 key를 식별자로 사용하면 확장성도 여러모로 떨어질 것 같네요!
반영했습니다 👍 

d252b258

yesjuhee commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -11,25 +11,36 @@
 
-@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
-@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
```
이번 미션을 하면서 기존의 `DEFINED_PORT` 설정을 `RANDOM_PORT`로 변경했습니다.
`RANDOM_PORT`를 설정하면 어플리케이션을 실행시키면서 동시에 테스트를 실행할 수 있다는 장점이 있습니다!
그리고 딱히 단점이 떠오르지 않아서 바꿨습니다. 혹시 `RANDOM_PORT`를 설정했을 때 예상할 수 있는 단점이 있을까요?

ddaaac replied:
잘 바꿔주셨습니다. 혹시라도 다른 어플리케이션이 켜져있는 경우에도 테스트가 돌아갈 수 있겠네요.

yesjuhee commented on src/test/java/roomescape/TestClientConfig.java:
```diff
@@ -0,0 +1,30 @@
+        return MockRestServiceServer
+                .bindTo(TEST_BUILDER)
+                .build();
+    }
```
테스트 환경에서 `MockRestServiceServer`를 설정하고, `TossPaymentsClient`의 요청을 `MockRestServiceServer`가 처리하도록 하기 위해서 `TestClientConfig`를 추가로 작성했습니다.

제가 파악한 `MockRestServiceServer`의 작동 방식은 아래와 같습니다. (시간이 없어서 내부를 뜯어보지는 못했고, 다른 크루의 도움을 받았습니다!)
1. `RestClient.Builder`를 `bindTo()`의 인자로 넘긴다.
2. `MockRestServiceServer` 내부에서 `RestClient`에 자기가 사용할 인터셉터를 등록한다.
3. 이렇게 빌드된 `RestClient`의 경우 요청을 보내면 `MockRestServiceServer`를 사용한다.

즉 테스트에서 `MockRestServiceServer`를 사용하기 위해서는 `RestClient.Builder`를 `MockRestServiceServer`에 등록해서 빌드 한 후, 해당 빌더를 재사용해 `RestClient`를 만들고 `TossPaymentsClient`에 넘겨주어서 `MockRestServiceServer`와 `TossPaymentsClient`가 같은 `RestClient`를 사용하도록 해야합니다.

따라서 프로덕션 코드의 Configuration을 사용할 수 없어서 이와 같은 방식을 선택했습니다!
혹시 이와 관련해서 개선점이나 더 좋은 방식이 있는지 궁금합니다!

ddaaac replied:
잘 구현해습니다.

API URL도 프로퍼티로 관리하면 이 값도 주입받아서 사용할 수 있을 것 같네요

yesjuhee replied:
반영했습니다!

223284a3

ddaaac commented on src/main/java/roomescape/client/TossPaymentsClient.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private void handlerTossPaymentsException(ClientHttpResponse res) throws IOException {
+        final ObjectMapper objectMapper = new ObjectMapper().configure(
```
스프링에서는 ObjectMapper도 빈으로 관리됩니다. 이 것도 주입받는 식으로 구현하면 어플리케이션 전반적으로 사용하는 ObjectMapper를 통일할 수 있을 것 같습니다. 추후에 ObjectMapper에 이런저런 설정이 들어갈 수도 있거든요.

yesjuhee replied:
8ce630bb

오호 ObjectMapper가 빈으로 관리된다는 것 새롭게 알아갑니다ㅎㅎ
반영했습니다!

ddaaac commented on src/main/java/roomescape/client/config/ClientConfig.java:
```diff
@@ -0,0 +1,37 @@
+
+    private final String secretKey;
+
+    public ClientConfig(@Value("${toss.payments.secret-key}") String secretKey) {
```
키를 프로퍼티로 관리하고 있군요 👍 
이렇게 구성하면 어떤 장점이 있을까요?

yesjuhee replied:
보안 측면에서 유용함이 있습니다. secret-key는 외부에 노출되면 안되는 값이기 때문에 내부 코드에 작성해서 사용하는 것은 매우 위험합니다. 키를 프로퍼티로 관리하고 프로퍼티 파일은 외부에 공유하지 않는 정책을 사용할 수 있습니다. 다만 이번 미션의 경우 키가 노출되어도 위험한 상황은 아니기 때문에 프로퍼티 파일을 프라이빗으로 다루지는 않았습니다!

또한 환경에 따라 달라질 수 있는 시크릿 키를 외부에서 쉽게 주입시킬 수 있다는 것도 장점입니다.

ddaaac commented on src/main/java/roomescape/client/config/ClientConfig.java:
```diff
@@ -0,0 +1,37 @@
+                RestClient.builder()
+                        .baseUrl("https://api.tosspayments.com/v1/payments/")
+                        .requestFactory(requestFactory)
+                        .defaultHeader("Authorization", getBasicAuthorizationValue())
```
HttpHeaders에 정의된 상수를 사용해볼 수 있을 것 같아요.

yesjuhee replied:
b91411da

이 부분도 새롭게 배워가네요 감사합니다 🫡

ddaaac commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -52,6 +70,15 @@ public static Reservation register(final Member member, final LocalDate date,
+    public static Reservation register(final Member member, final LocalDate date, final ReservationTime time,
+                                       final Theme theme, final Payment payment) {
+        Reservation reservation = new Reservation(null, member, date, time, theme, payment);
+        if (reservation.isBefore(LocalDateTime.now())) {
```
`now()`를 사용할 때는 꼭 타임존을 명시해주세요.
실행되는 머신에 따라서 의도치 않은 결과가 나올 수 있습니다.

yesjuhee replied:
오 좋은 포인트 감사합니다! 중요한 부분인데 지금까지 고려하지 못하고 있었네요

d842c819

ddaaac commented on src/test/java/roomescape/fixture/ServerClientFixture.java:
```diff
@@ -0,0 +1,20 @@
+            .bindTo(TEST_BUILDER)
+            .build();
+
+    public static final ObjectMapper MAPPER = new ObjectMapper();
```
앞에 config랑 겹치는 거 같은데 한 쪽만 유지하는 건 어떨까요?

yesjuhee replied:
4cb22d35

`ServerClientFixture`를 삭제하는 방향으로 리팩터링 했습니다. TossPaymentClient의 경우 스프링 부트를 사용하지 않는 단위 테스트이기 때문에 따로 Config 클래스를 사용하지 않고 테스트 클래스에서 직접 설정해주는 방식을 사용했습니다.

또한 모든 테스트에서 사용되는 BASE_URL도 TestClientConfig 의 변수를 public으로 여는 방식 대신 각 테스트 클래스에서 private 상수로 정의하여 사용하도록 했습니다. 상수도 가능하면 public 대신 private으로 사용해야 테스트의 응집도가 높아진다고 판단했기 때문입니다.

ddaaac reviewed (changes requested):
안녕하세요 노랑 👋 

구현 잘 해주셨네요.

몇 가지 코멘트 남겼으니 확인해주세요.

yesjuhee committed:
refactor: Toss API 에러 응답 처리 방식 변경

- `code` 필드를 `TossErrorResponse`에 추가
- 공식 문서를 참고해 Read Timeout 시간 변경

yesjuhee committed:
feat: 결제 확인 시 특정 예외 처리 로직 추가

- Toss API 에러 코드 중 특정 코드에 대해 사용자 메시지를 일괄 처리하도록 변경
- `TossErrorResponse` 생성자에 예외 처리 로직 추가
- 관련 테스트 케이스 추가

yesjuhee committed:
refactor: `Payment` 엔티티 식별자 변경

- `@GeneratedValue` 어노테이션 추가로 `Payment` 엔티티의 `id` 자동 생성 처리
- PaymentRepository 기본 키 타입 변경 (PaymentKey -> Long)

yesjuhee committed:
refactor: TOSS URL을 프로퍼티로 관리하도록 변경

- Toss Payments `baseUrl`을 설정 파일에서 주입받도록 변경
- `application.properties`를 `application.yml`로 변환
- `ClientConfig`에 `baseUrl` 필드 및 관련 생성자 매개변수 추가

yesjuhee committed:
refactor: `TossPaymentsClient`에 `ObjectMapper` 주입 방식 적용

yesjuhee committed:
refactor: `HttpHeaders` 상수 사용으로 코드 가독성 개선

yesjuhee committed:
refactor: `Reservation` 생성 시점 검증에 `ZoneDateTime` 사용 및 생성 로직 리팩터링

- `ZonedDateTime` 사용으로 시간대 고려한 예약 검증 추가
- `register` 메서드 오버로딩으로 코드를 간결하게 정리

yesjuhee committed:
refactor: `ServerClientFixture` 제거 및 관련 테스트 수정

- `ServerClientFixture` 사용 부분을 각 테스트 클래스에서 직접 정의하도록 수정
- 테스트 코드 내 중복 제거 및 가독성 개선
- `MAPPER` 및 `BASE_URL` 상수를 각 테스트 클래스에 통합하여 관리

ddaaac reviewed (approved):
안녕하세요 노랑 👋 

리뷰 반영 잘 해주셨네요.

다음 단계 진행해주세요!
